File: Algorithms/76. Minimum Window Substring/MinWindow.java
//Author: Tushar Jaiswal
//Creation Date: 10/24/2020

/*Given a string S and a string T, find the minimum window in S which will contain all the characters in T in complexity O(n).

Example:
Input: S = "ADOBECODEBANC", T = "ABC"
Output: "BANC"

Note:
* If there is no such window in S that covers all characters in T, return the empty string "".
* If there is such window, you are guaranteed that there will always be only one unique minimum window in S.*/

/*Runtime Complexity: O(number of characters in s)
Space Complexity: O(number of characters in t)*/

class Solution {
    public String minWindow(String s, String t) {
        if (s == null || t == null || s.equals("") || t.equals("")) {
            return "";
        }

        // Using Streams
        // Map<Character, Integer> charFreq = t.chars()
        //        .mapToObj(c -> (char)c)
        //        .collect(Collectors.groupingBy(Function.identity(), Collectors.summingInt(c -> 1)));
        HashMap<Character, Integer> charFreq = new HashMap<>();
        for (char c : t.toCharArray()) {
            charFreq.merge(c, 1, Integer::sum);
        }
        HashMap<Character, Integer> charsToMatch = new HashMap<>(charFreq);
        HashMap<Character, Integer> excessChars = new HashMap<>();

        String minWindow = null;
        int left = -1;
        int right = -1;

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);

            if(charFreq.containsKey(c)) {
                left = left == -1 ? i : left;

                if (charsToMatch.containsKey(c)) {
                    charsToMatch.replace(c, charsToMatch.get(c) - 1);
                    if (charsToMatch.get(c) == 0) {
                        charsToMatch.remove(c);
                        if(charsToMatch.size() == 0) {
                            right = i;
                            minWindow = s.substring(left, right + 1);
                            break;
                        }
                    }
                } else {
                    excessChars.merge(c, 1, Integer::sum);
                }
            }
        }
        if (minWindow == null) {
            return "";
        }


        while (left < right && right < s.length()) {
            char c = s.charAt(left);

            if(charFreq.containsKey(c) && !excessChars.containsKey(c)) {
                char newChar = '0';
                do {
                    right++;
                    if (right < s.length()) {
                    newChar = s.charAt(right);

                        if (newChar == c) {
                            left++;
                        } else if (charFreq.containsKey(newChar)) {
                            excessChars.merge(newChar, 1, Integer::sum);
                        }
                    }
                } while (right < s.length() && newChar != c);
            } else if (charFreq.containsKey(c) && excessChars.containsKey(c)) {
                left++;
                minWindow = minWindow.length() > (right - left + 1) ? s.substring(left, right + 1) : minWindow;
                excessChars.replace(c, excessChars.get(c) - 1);
                if (excessChars.get(c) == 0) {
                    excessChars.remove(c);
                }
            } else {
                left++;
                minWindow = minWindow.length() > (right - left + 1) ? s.substring(left, right + 1) : minWindow;
            }
        }

        return minWindow;
    }
}

File: Algorithms/76. Minimum Window Substring/MinWindow.py
# Author: Tushar Jaiswal
# Creation Date: 2025-08-06

# Problem Source: https://leetcode.com/problems/minimum-window-substring

# Given two strings s and t of lengths m and n respectively, return the minimum window substring of s such that every character in t (including duplicates) is included in the window. If there is no such substring, return the empty string "".

# The testcases will be generated such that the answer is unique.

# Example 1:
# Input: s = "ADOBECODEBANC", t = "ABC"
# Output: "BANC"
# Explanation: The minimum window substring "BANC" includes 'A', 'B', and 'C' from string t.

# Example 2:
# Input: s = "a", t = "a"
# Output: "a"
# Explanation: The entire string s is the minimum window.

# Example 3:
# Input: s = "a", t = "aa"
# Output: ""
# Explanation: Both 'a's from t must be included in the window.
# Since the largest window of s only has one 'a', return empty string.

# Constraints:
# m == s.length
# n == t.length
# 1 <= m, n <= 10^5
# s and t consist of uppercase and lowercase English letters.

# Follow up: Could you find an algorithm that runs in O(m + n) time?

# Runtime complexity: O(m)
# Space complexity: O(n)

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(s) < len(t):
            return ""
        
        freq_map = Counter(t)
        counter = len(freq_map)
        begin = end = 0
        length = float('inf')

        while end < len(s):
            c = s[end]
            end += 1

            if c in freq_map:
                freq_map[c] -=1
                if freq_map[c] == 0:
                    counter -= 1

            while counter == 0:
                temp_c = s[begin]
                if temp_c in freq_map:
                    if freq_map[temp_c] == 0:
                        counter += 1
                    freq_map[temp_c] += 1

                new_length = end - begin
                if new_length < length:
                    length = new_length
                    head = begin
                begin += 1

        return "" if length == float('inf') else s[head:head + length]
